<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Prepare Summarized Experiment — prepareSummarizedExperiment • bcbioBase</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Prepare Summarized Experiment — prepareSummarizedExperiment" />

<meta property="og:description" content="This is a utility wrapper for SummarizedExperiment::SummarizedExperiment()
that provides automatic subsetting for row and column data." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioBase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioBase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Prepare Summarized Experiment</h1>
    <small>Source: <a href='https://github.com/hbc/bcbioBase/blob/master/R/prepareSummarizedExperiment.R'><code>R/prepareSummarizedExperiment.R</code></a></small>
    </div>

    
    <p>This is a utility wrapper for <code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>
that provides automatic subsetting for row and column data.</p>
    

    <pre class="usage"><span class='fu'>prepareSummarizedExperiment</span>(<span class='no'>assays</span>, <span class='no'>rowRanges</span>, <span class='no'>colData</span>, <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>isSpike</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assays</th>
      <td><p>List containing RNA-seq count matrices with matching
dimensions. Counts can be passed in either dense (<code>matrix</code>) or sparse
(<code>dgCMatrix</code>, <code>dgTMatrix</code>) format.</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code> describing assay rows. Must contain genomic
ranges. Can be left <code>NULL</code> if the genome is poorly annotated and/or ranges
aren't available from AnnotationHub.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p><code>DataFrame</code> <code>data.frame</code>, or <code>matrix</code> describing assay
columns.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p><em>Optional</em>. Metadata <code>list</code>.</p></td>
    </tr>
    <tr>
      <th>isSpike</th>
      <td><p>Character vector of spike-in sequence rownames.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>RangedSummarizedExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function also provides automatic metadata slotting of multiple useful
environment parameters:</p><ul>
<li><p><code>date</code>: Today's date, returned from <code>Sys.Date()</code>.</p></li>
<li><p><code>wd</code>: Working directory, returned from <code>getwd()</code>.</p></li>
<li><p><code>utilsSessionInfo</code>: [utils::sessionInfo().</p></li>
<li><p><code>devtoolsSessionInfo</code>: [sessioninfo::session_info()].</p></li>
</ul>
    <p>[utils::sessionInfo().</p><ul>
<li><p><code>devtoolsSessionInfo</code>: [sessioninfo::session_info()]: R:utils::sessionInfo().<!-- %0A-%20%60devtoolsSessionInfo%60:%20[sessioninfo::session_info() --></p></li>
</ul>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(
    <span class='st'>"EGFP"</span>,  <span class='co'># spike</span>
    <span class='st'>"gene_1"</span>,
    <span class='st'>"gene_2"</span>,
    <span class='st'>"gene_3"</span>
)
<span class='no'>samples</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(
    <span class='st'>"sample_1"</span>,
    <span class='st'>"sample_2"</span>,
    <span class='st'>"sample_3"</span>,
    <span class='st'>"sample_4"</span>
)
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(
    <span class='fu'>seq</span>(<span class='fl'>1L</span>:<span class='fl'>16L</span>),
    <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>4L</span>,
    <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>4L</span>,
    <span class='kw'>dimnames</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='no'>genes</span>, <span class='no'>samples</span>)
)
<span class='no'>assays</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>assay</span> <span class='kw'>=</span> <span class='no'>mat</span>)
<span class='co'># Leave out the unannotated EGFP spike-in</span>
<span class='no'>rowRanges</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GRanges</a></span>(
    <span class='kw'>seqnames</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"1"</span>, <span class='st'>"1"</span>, <span class='st'>"1"</span>),
    <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='fu'><a href='http://www.rdocumentation.org/packages/IRanges/topics/IRanges-constructor'>IRanges</a></span>(
        <span class='kw'>start</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>1L</span>, <span class='fl'>101L</span>, <span class='fl'>201L</span>),
        <span class='kw'>end</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>100L</span>, <span class='fl'>200L</span>, <span class='fl'>300L</span>)
    )
)
<span class='fu'>names</span>(<span class='no'>rowRanges</span>) <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"gene_1"</span>, <span class='st'>"gene_2"</span>, <span class='st'>"gene_3"</span>)
<span class='no'>colData</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(
    <span class='st'>"genotype"</span> <span class='kw'>=</span> <span class='fu'>c</span>(
        <span class='st'>"wildtype"</span>,
        <span class='st'>"wildtype"</span>,
        <span class='st'>"knockout"</span>,
        <span class='st'>"knockout"</span>
    ),
    <span class='st'>"age"</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>3L</span>, <span class='fl'>6L</span>, <span class='fl'>3L</span>, <span class='fl'>6L</span>),
    <span class='kw'>row.names</span> <span class='kw'>=</span> <span class='no'>samples</span>
)
<span class='fu'>prepareSummarizedExperiment</span>(
    <span class='kw'>assays</span> <span class='kw'>=</span> <span class='no'>assays</span>,
    <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='no'>rowRanges</span>,
    <span class='kw'>colData</span> <span class='kw'>=</span> <span class='no'>colData</span>,
    <span class='kw'>isSpike</span> <span class='kw'>=</span> <span class='st'>"EGFP"</span>
)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 4 4 
#&gt; metadata(6): date wd ... isSpike unannotatedRows
#&gt; assays(1): assay
#&gt; rownames(4): EGFP gene_1 gene_2 gene_3
#&gt; rowData names(0):
#&gt; colnames(4): sample_1 sample_2 sample_3 sample_4
#&gt; colData names(2): genotype age</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

