<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read Sample Metadata — readSampleData • bcbioBase</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Read Sample Metadata — readSampleData" />

<meta property="og:description" content="This function reads user-defined sample metadata saved in a spreadsheet.
The &quot;description&quot; column is always required, and must match the bcbio
per sample directory names exactly. Inclusion of the &quot;fileName&quot; column
isn't required but is recommended for data provenance." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioBase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioBase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read Sample Metadata</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioBase/blob/master/R/readSampleData.R'><code>R/readSampleData.R</code></a></small>
    <div class="hidden name"><code>readSampleData.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function reads user-defined sample metadata saved in a spreadsheet.
The "<code>description</code>" column is always required, and must match the bcbio
per sample directory names exactly. Inclusion of the "<code>fileName</code>" column
isn't required but is recommended for data provenance.</p>
    
    </div>

    <pre class="usage"><span class='fu'>readSampleData</span>(<span class='no'>file</span>, <span class='kw'>lanes</span> <span class='kw'>=</span> <span class='fl'>1L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code>string</code>. File path. Supports CSV, TSV, and XLSX file formats.</p></td>
    </tr>
    <tr>
      <th>lanes</th>
      <td><p><code>scalar integer</code>. Number of lanes used to split the samples into
technical replicates (<code>_LXXX</code>) suffix.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>DataFrame</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Some bcbio examples on readthedocs use "<code>samplename</code>" instead of
"<code>fileName</code>". This function checks for that and will error out
intentionally, since we're using the <code>sampleName</code> column (note case) to
define unique sample names, in the event that bcbio has processed
multiplexed samples.</p>
    
    <h2 class="hasAnchor" id="demultiplexed-samples"><a class="anchor" href="#demultiplexed-samples"></a>Demultiplexed samples</h2>

    
    <p>This applies to bulk RNA-seq samples.</p>
<p>Normally when loading a bcbio run of demultiplexed samples, the sample
metadata will be imported automatically from the <code>project-summary.yaml</code> file
in the final upload directory using the <code><a href='readYAML.html'>readYAMLSampleData()</a></code> function. If
you notice any typos in your metadata after completing the run, these can be
corrected by editing the YAML file. Alternatively, you can pass in a
spreadsheet with the <code>readSampleData()</code> function.</p>
<p>The samples in the bcbio run must map to the <code>description</code> column. The values
provided in <code>description</code> for demultiplexed samples must be unique. They must
also be <em>syntactically valid</em>, meaning that they cannot contain illegal
characters (e.g. spaces, non-alphanumerics, <strong>dashes</strong>) or <strong>begin with a
number</strong>. Consult the documentation in <code>help(topic = "make.names")</code> for more
information on valid names in R.</p>
    
    <h2 class="hasAnchor" id="multiplexed-samples"><a class="anchor" href="#multiplexed-samples"></a>Multiplexed samples</h2>

    
    <p>This applies to some single-cell RNA-seq formats, including inDrops. In this
case, bcbio will output per-sample directories with this this structure:
<code>description</code>-<code>revcomp</code>.</p>
<p><code>readSampleData()</code> checks to see if the <code>description</code> column is unique. If
the values are duplicated, the function assumes that bcbio processed
multiplexed FASTQs, where multiple samples of interest are barcoded inside a
single FASTQ. This this case, you must supply additional "<code>index</code>",
"<code>sequence</code>", and "<code>sampleName</code>" columns.</p>
<p>Note that bcbio currently outputs the reverse complement index sequence in
the sample directory names (e.g. <code>sample-ATAGAGAG</code>). Define the forward index
barcode in the <code>sequence</code> column here, not the reverse complement. The
reverse complement will be calculated automatically and added as the
<code>revcomp</code> column in the sample metadata.</p>
<p>Here's an example of valid multiplexed sample data:</p>
<p>| description | index | sequence | sampleName | genotype |
|-------------|-------|----------|------------|----------|
| indrops     | 1     | CTCTCTAT | sample1    | wildtype |
| indrops     | 2     | TATCCTCT | sample2    | knockout |
| indrops     | 3     | GTAAGGAG | sample3    | wildtype |
| indrops     | 4     | ACTGCATA | sample4    | knockout |</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Read Functions: <code><a href='readDataVersions.html'>readDataVersions</a></code>,
  <code><a href='readLog.html'>readLog</a></code>, <code><a href='readProgramVersions.html'>readProgramVersions</a></code>,
  <code><a href='readTx2gene.html'>readTx2gene</a></code>, <code><a href='readYAML.html'>readYAML</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Demultiplexed</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='st'>"http://bcbiobase.seq.cloud/demultiplexed.csv"</span>
<span class='kw pkg'>readr</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/readr/topics/read_delim'>read_csv</a></span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  fileName = col_character(),</span>
#&gt; <span class='message'>  description = col_character(),</span>
#&gt; <span class='message'>  genotype = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; # A tibble: 4 x 3
#&gt;   fileName            description genotype
#&gt;   &lt;chr&gt;               &lt;chr&gt;       &lt;chr&gt;   
#&gt; 1 sample1_R1.fastq.gz sample1     wildtype
#&gt; 2 sample2_R1.fastq.gz sample2     knockout
#&gt; 3 sample3_R1.fastq.gz sample3     wildtype
#&gt; 4 sample4_R1.fastq.gz sample4     knockout</div><div class='input'><span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>readSampleData</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Reading demultiplexed.csv</span></div><div class='output co'>#&gt; <span class='message'>Demultiplexed samples detected</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; DataFrame with 4 rows and 4 columns
#&gt;         sampleName            fileName description genotype
#&gt;           &lt;factor&gt;            &lt;factor&gt;    &lt;factor&gt; &lt;factor&gt;
#&gt; sample1    sample1 sample1_R1.fastq.gz     sample1 wildtype
#&gt; sample2    sample2 sample2_R1.fastq.gz     sample2 knockout
#&gt; sample3    sample3 sample3_R1.fastq.gz     sample3 wildtype
#&gt; sample4    sample4 sample4_R1.fastq.gz     sample4 knockout</div><div class='input'>
<span class='co'># Multiplexed</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='st'>"http://bcbiobase.seq.cloud/multiplexed.csv"</span>
<span class='kw pkg'>readr</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/readr/topics/read_delim'>read_csv</a></span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  fileName = col_character(),</span>
#&gt; <span class='message'>  description = col_character(),</span>
#&gt; <span class='message'>  index = col_integer(),</span>
#&gt; <span class='message'>  sequence = col_character(),</span>
#&gt; <span class='message'>  sampleName = col_character(),</span>
#&gt; <span class='message'>  aggregate = col_character()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; # A tibble: 8 x 6
#&gt;   fileName             description index sequence sampleName aggregate
#&gt;   &lt;chr&gt;                &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;    
#&gt; 1 indrops1_R1.fastq.gz indrops1        1 CTCTCTAT sample1_1  sample1  
#&gt; 2 indrops1_R1.fastq.gz indrops1        2 TATCCTCT sample2_1  sample2  
#&gt; 3 indrops1_R1.fastq.gz indrops1        3 GTAAGGAG sample3_1  sample3  
#&gt; 4 indrops1_R1.fastq.gz indrops1        4 ACTGCATA sample4_1  sample4  
#&gt; 5 indrops2_R1.fastq.gz indrops2        1 CTCTCTAT sample1_2  sample1  
#&gt; 6 indrops2_R1.fastq.gz indrops2        2 TATCCTCT sample2_2  sample2  
#&gt; 7 indrops2_R1.fastq.gz indrops2        3 GTAAGGAG sample3_2  sample3  
#&gt; 8 indrops2_R1.fastq.gz indrops2        4 ACTGCATA sample4_2  sample4  </div><div class='input'><span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>readSampleData</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Reading multiplexed.csv</span></div><div class='output co'>#&gt; <span class='message'>Multiplexed samples detected</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; DataFrame with 8 rows and 7 columns
#&gt;                   sampleName             fileName       description    index
#&gt;                     &lt;factor&gt;             &lt;factor&gt;          &lt;factor&gt; &lt;factor&gt;
#&gt; indrops1_AGAGGATA  sample2_1 indrops1_R1.fastq.gz indrops1-AGAGGATA        2
#&gt; indrops1_ATAGAGAG  sample1_1 indrops1_R1.fastq.gz indrops1-ATAGAGAG        1
#&gt; indrops1_CTCCTTAC  sample3_1 indrops1_R1.fastq.gz indrops1-CTCCTTAC        3
#&gt; indrops1_TATGCAGT  sample4_1 indrops1_R1.fastq.gz indrops1-TATGCAGT        4
#&gt; indrops2_AGAGGATA  sample2_2 indrops2_R1.fastq.gz indrops2-AGAGGATA        2
#&gt; indrops2_ATAGAGAG  sample1_2 indrops2_R1.fastq.gz indrops2-ATAGAGAG        1
#&gt; indrops2_CTCCTTAC  sample3_2 indrops2_R1.fastq.gz indrops2-CTCCTTAC        3
#&gt; indrops2_TATGCAGT  sample4_2 indrops2_R1.fastq.gz indrops2-TATGCAGT        4
#&gt;                   sequence aggregate  revcomp
#&gt;                   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;
#&gt; indrops1_AGAGGATA TATCCTCT   sample2 AGAGGATA
#&gt; indrops1_ATAGAGAG CTCTCTAT   sample1 ATAGAGAG
#&gt; indrops1_CTCCTTAC GTAAGGAG   sample3 CTCCTTAC
#&gt; indrops1_TATGCAGT ACTGCATA   sample4 TATGCAGT
#&gt; indrops2_AGAGGATA TATCCTCT   sample2 AGAGGATA
#&gt; indrops2_ATAGAGAG CTCTCTAT   sample1 ATAGAGAG
#&gt; indrops2_CTCCTTAC GTAAGGAG   sample3 CTCCTTAC
#&gt; indrops2_TATGCAGT ACTGCATA   sample4 TATGCAGT</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#demultiplexed-samples">Demultiplexed samples</a></li>

      <li><a href="#multiplexed-samples">Multiplexed samples</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

